### PR TITLE
User engagement tracking

### LINKED ISSUE #60: Basic Gamification - Tracking User Engagement
- [x] Set up system to update karma for user easily

These engagement factors go through a function to generate your karma:
- [x] submit a snippet: +0.1 karma
- [x] voting on a snippet: +0.1 karma
- [x] winning a round: +5.0 karma
~- [ ] viewing projects: 0.01 x user karma (but only up to 100 views)~
- [x] compose a starter project: +3.0 karma
- [x] complete a badge: +10.0 karma

Purpose of karma
- Votes of users without much karma is worth very little
- Power of a user's vote is # votes x karma
- If a user votes for multiple snippets in a round then the weight of their vote is divided proportionately by the number of votes the user makes

### PR DESCRIPTION
**Pull request by meganyyu**
closes #60

### CODE REVIEW

File: Prolific/Prolific/Utils/Persistence/DAO.h [hunk 1]
```diff
@@ -22,7 +22,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 /** Save a user's data to database. */
 - (void)saveUser:(User *)user
-      completion:(void(^)(NSString *userId, NSError *error))completion;
+      completion:(void(^)(NSError *error))completion;
```
Comment (rickyv-fb) [inline Prolific/Prolific/Utils/Persistence/DAO.h:25]: yeah makes sense to not provide user id here

File: Prolific/Prolific/Utils/User Engagement Manager/UserEngagementManager.h [hunk 2]
```diff
@@ -8,10 +8,24 @@
 
 #import <Foundation/Foundation.h>
 
+#import "User.h"
+
 NS_ASSUME_NONNULL_BEGIN
 
 @interface UserEngagementManager : NSObject
 
+typedef NS_ENUM(NSInteger, UserEngagementType) {
```
Comment (rickyv-fb) [inline Prolific/Prolific/Utils/User Engagement Manager/UserEngagementManager.h:17]: Enums! nice job. You even got the annoying pattern on putting the whole enum name in each case :)

File: Prolific/Prolific/Utils/User Engagement Manager/UserEngagementManager.m [hunk 3]
```diff
@@ -8,6 +8,32 @@
 
 #import "UserEngagementManager.h"
 
+#import "UserBuilder.h"
+
 @implementation UserEngagementManager
 
++ (NSDecimalNumber *)computeKarmaValueForUserEngagementType:(UserEngagementType)userEngagementType {
```
Comment (rickyv-fb) [inline Prolific/Prolific/Utils/User Engagement Manager/UserEngagementManager.m:28]: nice. looks good and no defaults

Comment (rickyv-fb) [reply Prolific/Prolific/Utils/User Engagement Manager/UserEngagementManager.m:28]: For the future, you'd probably end up putting this logic on the server so Android / web doesn't have to do this. For the sake of the internship this makes sense of course

File: Prolific/Prolific/Utils/User Engagement Manager/UserEngagementManager.m [hunk 4]
```diff
@@ -12,6 +12,19 @@
...
 @implementation UserEngagementManager
 
++ (User *)updateKarmaForUser:(User *)user
+               forEngagement:(UserEngagementType)userEngagementType {
+    NSDecimalNumber *const newKarma = [UserEngagementManager computeKarmaValueForUserEngagementType:userEngagementType];
+    User *const updatedUser = [[[[UserBuilder alloc] initWithUser:user]
+                                addKarma:newKarma]
+                               build];
+    
+    if (updatedUser) {
```
Comment (rickyv-fb) [inline Prolific/Prolific/Utils/User Engagement Manager/UserEngagementManager.m:22]: interesting. The add karma will / can return a nil user?

File: Prolific/Prolific/Utils/Navigation/NavigationManager.h [hunk 5]
```diff
@@ -37,7 +38,7 @@ NS_ASSUME_NONNULL_BEGIN
 /** Presents ComposeSnippetViewController and passes in necessary view controller properties. */
 + (void)presentComposeSnippetViewControllerForRound:(Round *)round
                                           projectId:(NSString *)projectId
-                               navigationController:(UINavigationController *)navController;
+                                 fromViewController:(UIViewController<ComposeSnippetViewControllerDelegate> *)fromViewController;
```
Comment (rickyv-fb) [inline Prolific/Prolific/Utils/Navigation/NavigationManager.h:41]: perfect

File: Prolific/Prolific/View Controllers/Submissions View Controller/SubmissionsViewController.m [hunk 6]
```diff
@@ -132,11 +133,22 @@ - (void)didVote:(Snippet *)snippet {
                 __strong typeof (weakSelf) strongSelf = weakSelf;
                 if (strongSelf == nil) return;
                 
-                if (error) {
-                    NSLog(@"Error updating firebase with vote: %@", error.localizedDescription);
-                } else {
+                if (!error) {
                     strongSelf.round = round;
                     strongSelf.project = project;
+
```
Comment (rickyv-fb) [inline Prolific/Prolific/View Controllers/Submissions View Controller/SubmissionsViewController.m:141]: too many nslogs being left behind in commits

File: Prolific/Prolific/View Controllers/Profile View Controller/ProfileViewController.m [hunk 7]
```diff
@@ -44,20 +46,34 @@ - (void)viewDidLoad {
...
+    [self loadUser];
+    
     self.view.backgroundColor = [UIColor whiteColor];
     
     self.navigationItem.title = @"Profile";
     [super setupBackButton];
     
     [self setupCollectionView];
+    [_collectionView reloadData];
+    [_layout invalidateLayout];
```
Comment (rickyv-fb) [inline Prolific/Prolific/View Controllers/Profile View Controller/ProfileViewController.m]: Does this trigger a second collection view reload?

File: Prolific/Prolific/View Controllers/Submissions View Controller/SubmissionsViewController.m (hunk 6 above)
Comment (meganyyu) [reply Prolific/Prolific/View Controllers/Submissions View Controller/SubmissionsViewController.m:141]: sorry, I've been using them as placeholders, but I really do need to replace all of these with an error logger!

Comment (rickyv-fb) [reply Prolific/Prolific/View Controllers/Submissions View Controller/SubmissionsViewController.m:141]: No worries. Just trying to get you to build the habit  :)

**Pull request merged**